engine/rpcengine.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2014 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package engine
+
+import (
+	"strings"
+	"time"
+
+	"github.com/coreos/fleet/log"
+	"github.com/coreos/fleet/machine"
+	"github.com/coreos/fleet/pkg/lease"
+	"github.com/coreos/fleet/registry"
+)
+
+func (e *Engine) rpcLeadership(leaseTTL time.Duration, machID string) lease.Lease {
```

Comment:
have you looked at https://github.com/coreos/etcd/blob/master/clientv3/concurrency/election.go?